domain/access/service/service.go

```diff
@@ -119,10 +121,20 @@ type UserState interface {
 	// satisfies accesserrors.UserNotFound.
 	DisableUserAuthentication(context.Context, string) error
 
-	// UpdateLastLogin will update the last login time for the user.
-	// If no user is found for the supplied user name an error is returned that
-	// satisfies accesserrors.UserNotFound.
-	UpdateLastLogin(context.Context, string) error
+	// UpdateLastModelLogin will update the last login time for the user.
+	// The following error types are possible from this function:
+	// - accesserrors.UserNameNotValid: When the username is not valid.
+	// - accesserrors.UserNotFound: When the user cannot be found.
+	// - modelerrors.NotFound: If no model by the given modelUUID exists.
+	UpdateLastModelLogin(context.Context, string, coremodel.UUID) error
+
+	// LastModelLogin will return the last login time of the specified user.
+	// The following error types are possible from this function:
+	// - accesserrors.UserNameNotValid: When the username is not valid.
+	// - accesserrors.UserNotFound: When the user cannot be found.
+	// - accesserrors.UserNeverConnectedToModel: If there is no record of the user
+	// connecting to the model.
+	LastModelLogin(context.Context, string, coremodel.UUID) (time.Time, error)
```

Comment:
There should probably exist a `modelerrors.NotFound` in this list for `LastModelLogin` ?

Reply:
Currently the error returned in that case would be `UserNeverAccessedModel`. I've just changed it now though to do a check for a missing model in the case that it gets an ErrorNoRows from the database so that we return the correct error in that case. I've added it now to this comment too.